### PR TITLE
[gen4 planner] Move more horizon planning to operators

### PR DESCRIPTION
**Pull request by systay**
## Description
While planning, we use two important types of data structures - logical plans and operators. We are moving more and more logic over to the operators. This PR moves planning of projections over to the operators. 

Still to do is: ordering, limit, distinct and aggregation.

## Related Issue(s)

Tracking issue: https://github.com/vitessio/vitess/issues/11626

Other PRs in this series: #11622, #11680, #12506

### CODE REVIEW

File: go/vt/vtgate/planbuilder/operator_transformers.go [hunk 1]
```diff
@@ -59,18 +59,36 @@ func transformToLogicalPlan(ctx *plancontext.PlanningContext, op ops.Operator, i
...
+		if err != nil {
+			return nil, err
+		}
+
+		return &simpleProjection{
+			logicalPlanCommon: newBuilderCommon(src),
+			eSimpleProj: &engine.SimpleProjection{
+				Cols: op.Columns,
+			},
+		}, nil
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operator_transformers.go:73]: Let's extract this to its own function so we can have a one line `return` in each `case`.

File: go/vt/vtgate/planbuilder/operator_transformers.go [hunk 2]
```diff
@@ -59,18 +59,36 @@ func transformToLogicalPlan(ctx *plancontext.PlanningContext, op ops.Operator, i
...
+		// this might already have been done on the operators
+		if predicate == nil {
+			predicate, err = evalengine.Translate(ast, &evalengine.Config{
+				ResolveColumn: resolveFromPlan(ctx, plan, true),
+				Collation:     ctx.SemTable.Collation,
+			})
+			if err != nil {
+				return nil, err
+			}
 		}
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operator_transformers.go:93]: Same thing as above here. Let's move this to its own extracted method.

File: go/vt/vtgate/planbuilder/operators/apply_join.go [hunk 3]
```diff
@@ -140,54 +140,80 @@ func (a *ApplyJoin) AddJoinPredicate(ctx *plancontext.PlanningContext, expr sqlp
...
-func (a *ApplyJoin) AddColumn(ctx *plancontext.PlanningContext, expr sqlparser.Expr) (int, error) {
+func (a *ApplyJoin) pushColLeft(ctx *plancontext.PlanningContext, e *sqlparser.AliasedExpr) (int, error) {
+	newLHS, offset, err := a.LHS.AddColumn(ctx, e, true)
+	if err != nil {
+		return 0, err
+	}
+	a.LHS = newLHS
+	return offset, nil
+}
+func (a *ApplyJoin) pushColRight(ctx *plancontext.PlanningContext, e *sqlparser.AliasedExpr) (int, error) {
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/apply_join.go:151]: Let's add a new line here

```suggestion
}

func (a *ApplyJoin) pushColRight(ctx *plancontext.PlanningContext, e *sqlparser.AliasedExpr) (int, error) {
```

File: go/vt/vtgate/planbuilder/operators/derived.go [hunk 4]
```diff
@@ -132,28 +134,41 @@ func (d *Derived) AddPredicate(ctx *plancontext.PlanningContext, expr sqlparser.
...
+		return nil, 0, vterrors.VT13001("cannot push non-colname expression to a derived table")
+	}
+
+	if reuseCol {
+		for offset, column := range d.Columns {
+			if ctx.SemTable.EqualsExpr(col, column) {
+				return d, offset, nil
+			}
+		}
 	}
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/derived.go:149]: We're using the same small logic in `ApplyJoin.AddColumn`, how do you feel about extracting this to a common function?

```suggestion
	if offset, ok := myExtractedFindColFunc(ctx, reuseCol, d.Columns, col); ok {
		return d, offset, nil
	}
```

```go
func myExtractedFindColFunc(ctx, reuseCol, d.Columns, col) (int, bool) {
	if !reuseCol {
		return 0, false
	}

	for offset, column := range d.Columns {
		if ctx.SemTable.EqualsExpr(col, column) {
			return offset, true
		}
	}
	return 0, false
}
```

File: go/vt/vtgate/planbuilder/operators/horizon.go [hunk 5]
```diff
@@ -22,8 +22,12 @@ import (
 	"vitess.io/vitess/go/vt/vtgate/planbuilder/plancontext"
 )
 
-// Horizon is an operator we use until we decide how to handle the source to the horizon.
-// It contains information about the planning we have to do after deciding how we will send the query to the tablets.
+// Horizon is an operator that allows us to postpone planning things like SELECT/GROUP BY/ORDER BY/LIMIT until later.
+// It contains information about the planning we have to do after deciding how we will send the query to the tablets.	// If we are able to push down the Horizon under a route, we don't have to plan these things separately and can
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/horizon.go:26]: New line here
```suggestion
// It contains information about the planning we have to do after deciding how we will send the query to the tablets.
// If we are able to push down the Horizon under a route, we don't have to plan these things separately and can
```

File: go/vt/vtgate/planbuilder/operators/horizon_planning.go [hunk 6]
```diff
@@ -19,60 +19,196 @@ package operators
...
 		}
-	})
+	}
+
+	newOp, err := rewrite.BottomUp(root, TableID, visitor, stopAtRoute)
+	if err != nil {
+		if vterr, ok := err.(*vterrors.VitessError); ok && vterr.ID == "VT13001" {
+			// we encountered a bug. let's try to back out
+			return nil, errNotHorizonPlanned
+		}
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/horizon_planning.go:72]: nice

File: go/vt/vtgate/planbuilder/operators/horizon_planning.go [hunk 7]
```diff
@@ -19,60 +19,196 @@ package operators
...
 	}
 
 	needsOrdering := len(qp.OrderExprs) > 0
 	canShortcut := isRoute && sel.Having == nil && !needsOrdering
+	_, isDerived := in.Source.(*Derived)
 
-	if !qp.NeedsAggregation() && sel.Having == nil && canShortcut && !needsOrdering && !qp.NeedsDistinct() && in.Select.GetLimit() == nil {
+	switch {
+	case qp.NeedsAggregation() || sel.Having != nil || sel.Limit != nil || isDerived || needsOrdering || qp.NeedsDistinct():
+		return nil, errNotHorizonPlanned
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/horizon_planning.go:105]: so this is some of the cases we do not handle in the Horizon Planning Operator today?

File: go/vt/vtgate/planbuilder/operators/simpleProjection.go
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/simpleProjection.go:1]: Let's rename this file to `go/vt/vtgate/planbuilder/operators/simpleProjection.go` to be consistent with the other files in the package

File: go/vt/vtgate/planbuilder/operators/table.go [hunk 8]
```diff
@@ -85,15 +95,18 @@ func (to *Table) TablesUsed() []string {
...
 	}
-	cols := op.GetColumns()
-	for idx, column := range cols {
-		if col.Name.Equal(column.Name) {
-			return idx, nil
+	sqlparser.RemoveKeyspaceFromColName(col)
+	cols := op.GetColNames()
+	if reuseCol {
+		for idx, column := range cols {
+			if col.Name.Equal(column.Name) {
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/table.go:107]: Same here about that little snippet

**Review by frouioui — APPROVED** [review]
Left some comments and question, otherwise it looks good to me! This is a great progress 🙏🏻

File: go/vt/vtgate/planbuilder/operators/horizon_planning.go (hunk 7 above)
Comment (systay) [reply go/vt/vtgate/planbuilder/operators/horizon_planning.go:105]: exactly

File: go/vt/vtgate/semantics/semantic_state.go [hunk 9]
```diff
@@ -129,6 +129,16 @@ func (st *SemTable) CopyDependencies(from, to sqlparser.Expr) {
 	st.Direct[to] = st.DirectDeps(from)
 }
 
+// CopyDependencies copies the dependencies from one expression into the other
+func (st *SemTable) Cloned(from, to sqlparser.SQLNode) {
```
Comment (GuptaManan100) [inline go/vt/vtgate/semantics/semantic_state.go:133]: The name of the function doesn't match the comment.

File: go/vt/vtgate/planbuilder/operators/filter.go [hunk 10]
```diff
@@ -27,6 +28,10 @@ import (
 type Filter struct {
 	Source     ops.Operator
 	Predicates []sqlparser.Expr
+
+	// FinalPredicate is the evalengine expression that will finally be used.
+	// It contains the ANDed predicates in Predicates, with ColName:s replaced by Offset:s
```
Comment (GuptaManan100) [inline go/vt/vtgate/planbuilder/operators/filter.go:33]: What does `Colname:s` mean? Also `Offest:s`?

**Review by GuptaManan100 — APPROVED** [review]

**Pull request merged**